### DISCUSSION TITLE
Custom Domains for deployed apps

### DISCUSSION THREAD

**Discussion by ndom91** [Budibase/budibase#1742] [Help] [answered]
Hey yall, just wondering if theres a prefered way to tie a custom domain to a deployed application?

Of course you could just reverse proxy the generated URL through nginx or somethign similar, but is there a more "first-party" solution out there? For example, I want my `https://budibase.company.tld/app_edeb16ddc7754f628ea3cc78ac61aa79` to be available at `https://newapp.company.tld`

**Comment by mike12345567** [comment, marked as answer]
Hi @ndom91 you can currently use the apps name to direct to the app instead of the app ID, using something like `https://budibase.company.tld/app/<app_name>` - if you have any non-safe URI characters in your apps name then you'll need to swap these for the `encodeURIComponent` equivalent when directing to the app. In the future there will likely be a bit more control around this, like being able to set a specific URL path for the app, un-related to the apps name.

As for sub-domain/deeper domain integration, that isn't something that we currently support, for us to support something like that in an out of the box manner it would require us putting quite strict rules in place about how the domain should be directed towards Budibase, for a business with existing infrastructure in place this felt like a bad approach.

**Reply by ndom91** [reply]
First of all, thanks for the quick response!

So yeah, I can appreciate how supporting custom subdomains out of the box is a bigger issue. But, how about supporting some sort of CNAME redirect type of deal? 

You'd still only have to have the user setup 1 domain/subdomain for budibase, `budibase.company.com`, for example. Then have them point all their apps as CNAME's to that. 

So in my DNS settings i'd setup: `app1.company.com CNAME budibase.company.com`, then in envoy you continue to forward all traffic like that on to `app-service`, then in there in koa you can use some subdomain middleware like [`koa-sub-domain`](https://www.npmjs.com/package/koa-sub-domain) or something like that to route to the correct app. 

Idk just the first thing that comes to mind, might have some holes upon deeper inspection haha.

**Reply by mike12345567** [reply]
If there was interest in something like this then it is certainly something we could look like, nothing is impossible after all!

One issue with this however is that it sort of defeats a lot of the functionality we are trying to build with Budibase, specifically around having a centralised portal from which you can be directed to all of your apps, if you login to your Budibase platform with a non admin/builder account you'll be taken to the portal, which can be customised with company logo and name. The user is only able to see the apps they have access to in the portal as well, so it's nice centralised location to manage all of this stuff from!

**Reply by adecor-dev** [reply]
@mike12345567
Please clarify, if using budibase cloud I DO NOT have the ability to use custom domain. Rather, if I self-host then I DO have the option to use custom domain?

Also, if I use the self-hosted option I CAN'T use a custom domain for each app. The domain used will be used for all apps built under that domain?

**Reply by shogunpurple** [reply]
Hi @adecor-dev - thats correct. In our cloud your apps are hosted at `yourtenant.budibase.app`. If you self-host, you don't get custom domains by default - you will need to buy a domain, and associate it with a reverse proxy (details [here](https://docs.budibase.com/self-hosting/hosting-settings/advanced-options/reverse-proxy).

If you are self-hosted, you can use a custom domain for each app, again through reverse proxy configuration and just adding different subdomains for different apps. 

You can, of course set up a reverse proxy that you host or a CNAME record in your domain and have it point to our cloud, but self-hosting will certainly give you much more control.

**Reply by juusoautiosalo** [reply]
As someone currently trying out Budibase, I would see the proposition by ndom91 as a very important factor in deciding whether Budibase is useful for us. Being able to host an app directly from a subdomain would be a significantly more elegant way of deploying apps. E.g. GitHub Pages enables hosting from a custom domain and I find it extremely useful.

The centralized app portal sounds nice, but it is unnecessary in the beginning if you have only one application... And I think companies probably want to list the apps somewhere else anyway. And I think there will be still methods to create the portal even if you enable custom domains for every app.

P.S. This is just honest feedback from a potential user, of course you make your own decisions. Thank you for the software in any case!

**Reply by mjashanks** [reply]
Thanks @juusoautiosalo - I agree that this would be a useful feature - I've no doubt that we will implement this someday, but we just have other priorities right now. Thanks for the feedback :)

**Reply by sebastianjung** [reply]
I'im surprised how this is not prioritized. 

We as an agency can't create admin panels for our customers' websites that are only available under ..budibase domains. It just does not appear professional.

Any news on this maybe? Don't want to sound rude. Great piece of software for sure!

**Reply by jacobwl** [reply]
Completely agree with @sebastianjung. Not being able to set a custom domain is a non-starter for any partner/agency/external use cases.

Any update, @shogunpurple?

**Reply by shogunpurple** [reply]
@sebastianjung @jacobwl 

Nothing in roadmap just yet - planning on looking at this over the next few months. Just to note as well this will be a paid feature on our cloud platform.

**Reply by jacobwl** [reply]
Thank you for the quick reply, @shogunpurple. Much appreciated.

No issue with the feature being paid.

**Comment by dosandco** [comment]
Hello all, I just came across this as I had assumed (🤦🏽‍♂️) that it would be possible on BB cloud.
I absolutely love the BB software - we found it on ProductHunt and migrated a complex JetAdmin app across in a long afternoon, and it feels both more stable and more configurable immediately.
As an agency, we would love to host the client portal (agency.business.app) on our own domain (at dev.agencydomain.com), giving clients access to their apps in development and then when ready to roll them out with their users, we’d happily pay to have them on the client’s own domain with CNAME (app.clientdomain.com), or even upgrade to the next tier to enable it.
Has this found a little corner of the roadmap going forward? 🤞🏽🤞🏽🤞🏽
